cmd/crossplane/main.go

```diff
@@ -38,106 +40,166 @@ import (
-	)
-	cmd := kingpin.MustParse(app.Parse(os.Args[1:]))
-
-	zl := zap.New(zap.UseDevMode(*debug))
+	baseCmd := filepath.Base(os.Args[0])
+	args := os.Args[1:]
+
+	app := kingpin.New(baseCmd, "An open source multicloud control plane.").DefaultEnvars()
+	debug := app.Flag("debug", "Run with debug logging.").Short('d').Bool()
+	syncPeriod := app.Flag("sync", "Controller manager sync period duration such as 300ms, 1.5h or 2h45m").Short('s').Default("1h").Duration()
+
+	a := appReq{
+		app:        app,
+		zl:         zap.New(zap.UseDevMode(*debug)),
+		syncPeriod: syncPeriod,
+	}
+
+	commands := []fullCommand{}
+	commands = append(commands, a.crossplaneCmd(baseCmd)...)
+	commands = append(commands, a.stackCmd()...)
+
+	cmdGiven := kingpin.MustParse(app.Parse(args))
+
 	if *debug {
 		// The controller-runtime runs with a no-op logger by default. It is
 		// *very* verbose even at info level, so we only provide it a real
 		// logger when we're running in debug mode.
-		ctrl.SetLogger(zl)
+		ctrl.SetLogger(a.zl)
+	}
+
+	knownCmd := false
+	for _, cmd := range commands {
+		if cmdGiven == cmd.clause.FullCommand() {
+			knownCmd = true
+			cmd.handler()
+			break
+		}
+	}
+
+	if !knownCmd {
+		kingpin.FatalUsage("unknown command %s", cmdGiven)
 	}
+}
 
-	// TODO(negz): Is there a reason these distinct pieces of functionality
-	// should all be part of the same binary? Should we break them up?
-	switch cmd {
+// crossplaneCmd provides the fullCommand for the default command
+//
+// The default command runs the controllers for all Crossplane resources except
+// stack resources.
+func (a appReq) crossplaneCmd(baseCmd string) []fullCommand {
+	crossplaneCmd := a.app.Command(baseCmd, "Start core Crossplane controllers.").Default()
 
-	case crossplaneCmd.FullCommand():
-		log := logging.NewLogrLogger(zl.WithName("crossplane"))
-		log.Debug("Starting", "sync-period", syncPeriod.String())
+	return []fullCommand{{crossplaneCmd, func() {
+		log := logging.NewLogrLogger(a.zl.WithName("crossplane"))
+		log.Debug("Starting", "sync-period", a.syncPeriod.String())
 
 		cfg, err := ctrl.GetConfig()
 		kingpin.FatalIfError(err, "Cannot get config")
 
-		mgr, err := ctrl.NewManager(cfg, ctrl.Options{SyncPeriod: syncPeriod})
+		mgr, err := ctrl.NewManager(cfg, ctrl.Options{SyncPeriod: a.syncPeriod})
 		kingpin.FatalIfError(err, "Cannot create manager")
 
 		kingpin.FatalIfError(apis.AddToScheme(mgr.GetScheme()), "Cannot add core Crossplane APIs to scheme")
 		kingpin.FatalIfError(oam.Setup(mgr, log), "Cannot setup OAM controllers")
 		kingpin.FatalIfError(workload.Setup(mgr, log), "Cannot setup workload controllers")
 		kingpin.FatalIfError(mgr.Start(ctrl.SetupSignalHandler()), "Cannot start controller manager")
+	}}}
+}
+
+// stackCmd provides the fullCommand for the "stack" command
+func (a appReq) stackCmd() []fullCommand {
```

Comment:
Do we need this method? What's the advantage of bundling the stack commands together in a method in this way? As a potential future maintainer, I would personally prefer that they weren't bundled, since there are only two, and because this adds another level of indirection to following the full set of commands around to see how they're generated and bound.

Reply:
We've been adding lots of changes to `stackManageCmd` lately and I don't think that is going to stop. I kind of like that we don't have to get involved with the other commands while editing the stack manager.

The variables are kept in their function scope so they could be renamed to simplify them because there is no longer ambiguity. (remove `stackManager` prefix on variables in the `stackManageCmd`).

Reply:
My attempt to appeal to a higher authority doesn't seem to put much weight on either argument:

https://github.com/golang/go/wiki/CodeReviewComments#line-length

> This is, actually, exactly the same advice about how long a function should be. There's no rule "never have a function more than N lines long", but there is definitely such a thing as too long of a function, and of too stuttery tiny functions, and the solution is to change where the function boundaries are, not to start counting lines.

I was looking for something along the lines of "keep a function short and focused on one thing".